### DISCUSSION TITLE
Error: @vercel/kv: Missing required environment variables KV_REST_API_URL and KV_REST_API_TOKEN

### DISCUSSION THREAD

**Discussion by 24jr** [vercel/community#2301] [Help] [answered]
I created my kv store and just installed vercel cli for first time and followed basic guide 

It created my .env.development.local file with the filled-in values

when I add the `import kv from "@vercel/kv"` to  +page.server.js file i get `Error: @vercel/kv: Missing required environment variables KV_REST_API_URL and KV_REST_API_TOKEN` however if I add `import { KV_REST_API_URL, KV_REST_API_TOKEN } from "$env/static/private"` I can see the actual values show up.

Do i need to do something else, is this a bug, or should i just use rest way. I actually just added upstash last week ironically and got the basics working with that but now going to switch over to staying in vercel ecosystem.

even if i try
```
import { KV_REST_API_URL, KV_REST_API_TOKEN } from "$env/static/private"
import { createClient } from "@vercel/kv"

// const kv = createClient({
//   url: KV_REST_API_URL,
//   token: KV_REST_API_TOKEN,
// })
```
i get that error

**Comment by amyegan** [comment]
This sounds similar to https://github.com/vercel/community/discussions/2281. Are you getting this error in server-side code, or in client side code?

**Comment by amyegan** [comment]
It looks like you're both running into trouble adding Vercel KV to an existing project rather than a problem with one of the example templates. I'm going to convert this into a discussion in our [support community](https://github.com/vercel/community/discussions)

**Comment by amyegan** [comment, marked as answer]
Are you getting this error in production, or only in local dev?

😕 I have had success with starting a new project from the [kv-redis-sveltekit](https://github.com/vercel/examples/tree/main/storage/kv-redis-sveltekit) starter and with adding Vercel KV to an existing project using the steps in the [Vercel KV Quickstart](https://vercel.com/docs/storage/vercel-kv/quickstart#) guide. There are a few steps that could easily be missed:
- [select all environments](https://github.com/orgs/vercel/discussions/2281#discussioncomment-5795368) where the store should be available
- for local development, [pull the related environment variables](https://vercel.com/docs/storage/vercel-kv/quickstart#preparing-your-local-project)
- make sure your `kv` code is server side, not running on the client

Can you walk me through the steps you tried, or share a link to the guide you used?

**Reply by amyegan** [reply]
I reached out to the storage engineers and they shared another issue reporting the same problem: https://github.com/vercel/storage/issues/107 

They're working on a solution now. 

As a workaround, you could try using [`vercel dev`](https://vercel.com/docs/cli/dev#) instead of your typical local dev command. I was able to reproduce the error in my test project using `npm run dev`, but it worked for me with `vercel dev`.

**Reply by NathanMLu** [reply]
Thank you for the update. I am facing the same issue, and was able to use the workaround you mentioned with `vercel dev` for a temporary fix.

**Reply by nicodes** [reply]
@amyegan Thank you this did the trick! I'm currently using this `kv` call in `hooks.server.ts`. I was wondering why `vercel dev` is needed, can it added to the docs?

**Comment by marfalkov** [comment]
@24jr @amyegan 

I ran into the same problem in local development with  [serverless-redis-http](https://github.com/hiett/serverless-redis-http)
Exporting the variables from the `.env` file solved this issue for me:

```bash
set -a # automatically export all variables
source .env
set +a
```

Then I was able to use `npm run dev`

**Reply by paulcalcraft** [reply]
If you have any # comment lines in your .env, this breaks (silently!)

Instead you can do this:
```bash
export $(cat .env | sed 's/#.*//g' | xargs)
```

**Comment by firatkucuk** [comment]
This problem still exists and `vercel dev` solves issue partially, then I get a CORS error. @marfalkov's method did work.

**Comment by assetmantledevops** [comment]
facing the same issues. also not all code can be server side right? when we are storing data to the kv store, it comes from client side since cliend adds the data using input textbox. It has to be in client side. so how to implement kv in client side ?

**Comment by jackprogramsjp** [comment]
Still, facing this issue. Uncommenting my .env file worked, but this has been mentioned since last year, really need to fix this really tiny bug.

**Reply by ghost** [reply]
I can confirm that it is the `#comments` in the `.env` file that are causing this error. 😂

**Comment by aqzi** [comment]
I had this issue when testing a few things on client side with nextjs, vercel and redis.

Solution:
1. Move the code to server side
2. When keeping code client side, declare env variables in the next.config.js file as follow:

``` typescript
const nextConfig = {
    env: {
        KV_URL: process.env.KV_URL,
        KV_REST_API_URL: process.env.KV_REST_API_URL,
        KV_REST_API_TOKEN: process.env.KV_REST_API_TOKEN,
        KV_REST_API_READ_ONLY_TOKEN: process.env.KV_REST_API_READ_ONLY_TOKEN,
    }
};
```

**Reply by oceans404** [reply]
This worked for me ^

**Reply by lazizia** [reply]
this worked for me too ^^^ thanks
